### PR TITLE
8335159: Move method reference to lambda desugaring before Lower

### PR DESCRIPTION
**Pull request by mcimadamore**
As we recently [moved](https://github.com/openjdk/jdk/pull/19836) the translation of "simple" method references from `LambdaToMethod` to `Lower`, it became clearer that this step in fact would benefit from running even *earlier*, as it depends on the synthetic casts generated by `TransTypes` (e.g. if one or more desugared lambda parameters have a type that is either an intersection or a union type).
Moving the translation earlier would allow us not to *guess* which casts would need to be introduced, and just running `TransTypes::translate` on the desugaring code would take into account the type mismatches.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issues
 * [JDK-8335159](https://bugs.openjdk.org/browse/JDK-8335159): Move method reference to lambda desugaring before Lower (**Enhancement** - P4)
 * [JDK-8336320](https://bugs.openjdk.org/browse/JDK-8336320): NullPointerException: Cannot invoke Type.getTag because type is null after JDK-8334037 (**Bug** - P2)

### Reviewers
 * [Vicente Romero](https://openjdk.org/census#vromero) (@vicente-romero-oracle - **Reviewer**) 🔄 Re-review required (review applies to [9ef90711](https://git.openjdk.org/jdk/pull/19903/files/9ef90711bc32392ca69bda51186e8639ef2c34b1))
 * [Jan Lahoda](https://openjdk.org/census#jlahoda) (@lahodaj - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19903/head:pull/19903` \
`$ git checkout pull/19903`

Update a local copy of the PR: \
`$ git checkout pull/19903` \
`$ git pull https://git.openjdk.org/jdk.git pull/19903/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19903`

View PR using the GUI difftool: \
`$ git pr show -t 19903`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19903.diff">https://git.openjdk.org/jdk/pull/19903.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19903#issuecomment-2191423896)

### CODE REVIEW

File: src/jdk.compiler/share/classes/com/sun/tools/javac/comp/TransTypes.java [hunk 1]
```diff
@@ -539,6 +546,270 @@ public void visitLambda(JCLambda tree) {
...
     }
 
+    @Override
+    public void visitReference(JCMemberReference tree) {
+        if (needsConversionToLambda(tree)) {
+            // Convert to a lambda, and process as such
+            MemberReferenceToLambda conv = new MemberReferenceToLambda(tree);
+            result = translate(conv.lambda());
+        } else {
+            Type t = types.skipTypeVars(tree.expr.type, false);
```
Comment (mcimadamore) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/comp/TransTypes.java:556]: This part is just the old `visitReference`

File: src/jdk.compiler/share/classes/com/sun/tools/javac/comp/TransTypes.java [hunk 2]
```diff
@@ -539,6 +546,270 @@ public void visitLambda(JCLambda tree) {
...
+                !receiverAccessible(tree) ||
+                (tree.getMode() == ReferenceMode.NEW &&
+                        tree.kind != ReferenceKind.ARRAY_CTOR &&
+                        (tree.sym.owner.isDirectlyOrIndirectlyLocal() || tree.sym.owner.isInner()));
+    }
+
+    /**
+     * Converts a method reference which cannot be used directly into a lambda
+     */
+    private class MemberReferenceToLambda {
```
Comment (mcimadamore) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/comp/TransTypes.java:657]: I've considered adding a completely new step for this, but decided against it, as for now there is an assumption that `TransTypes` is the first translation step after `Flow`, and we run a scanner before it to collect usages of various features, as well as supertype/subtype ordering information (as `TransTypes` needs to be run in a certain order - e.g. supertypes first).

File: src/jdk.compiler/share/classes/com/sun/tools/javac/comp/TransTypes.java [hunk 3]
```diff
@@ -539,6 +546,270 @@ public void visitLambda(JCLambda tree) {
...
+    }
+
+    /**
+     * This method should be called only when target release <= 14
+     * where LambdaMetaFactory does not spin nestmate classes.
+     *
+     * This method should be removed when --release 14 is not supported.
+     */
+    boolean isPrivateInOtherClass(JCMemberReference tree) {
+        assert !target.runtimeUseNestAccess();
```
Comment (vicente-romero-oracle) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/comp/TransTypes.java:630]: nit: this assertion can be removed

**Review by vicente-romero-oracle — APPROVED** [review]
looks good

**Review by vicente-romero-oracle — APPROVED** [review]

**Review by lahodaj — APPROVED** [review]
Looks OK to me.

File: src/jdk.compiler/share/classes/com/sun/tools/javac/comp/TransTypes.java [hunk 4]
```diff
@@ -539,6 +546,268 @@ public void visitLambda(JCLambda tree) {
...
+    }
+
+    /**
+     * This method should be called only when target release <= 14
+     * where LambdaMetaFactory does not spin nestmate classes.
+     *
+     * This method should be removed when --release 14 is not supported.
+     */
+    boolean isPrivateInOtherClass(JCMemberReference tree) {
+        return  (tree.sym.flags() & PRIVATE) != 0 &&
```
Comment (turbanoff) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/comp/TransTypes.java:629]:
```suggestion
        return (tree.sym.flags() & PRIVATE) != 0 &&
```

File: test/langtools/tools/javac/SuperInit/MrefDoubleTrans.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+ * @summary NullPointerException: Cannot invoke Type.getTag because type is null after JDK-8334037
+ * @compile MrefDoubleTrans.java
+ */
+class MrefDoubleTrans {
+    static class T {
+        public void f() {
+            Runnable r = new I()::m;
+        }
+
+        static class I {
```
Comment (lahodaj) [inline test/langtools/tools/javac/SuperInit/MrefDoubleTrans.java:36]: To make the test fail without this patch:
```suggestion
        class I {
```

**Review by lahodaj — APPROVED** [review]
Looks good!

**Pull request closed without merging**